src/Nethermind/Nethermind.AuRa/AuRaBlockFinalizationManager.cs

```diff
@@ -0,0 +1,269 @@
+                LastFinalizedBlockLevel = finalizedBlocks[finalizedBlocks.Count - 1].Number;
+            }
+        }
+        
+        private IReadOnlyList<BlockHeader> GetFinalizedBlocks(BlockHeader block)
+        {
+            (ChainLevelInfo parentLevel, BlockInfo parentBlockInfo) GetBlockInfo(BlockHeader blockHeader)
+            {
+                var chainLevelInfo = _blockInfoRepository.LoadLevel(blockHeader.Number);
+                var blockInfo = chainLevelInfo.BlockInfos.First(i => i.BlockHash == blockHeader.Hash);
+                return (chainLevelInfo, blockInfo);
+            }
+            
+            var minSealersForFinalization = block.IsGenesis ? 1 : _auRaValidator.MinSealersForFinalization;
+            var originalBlock = block;
+            
+            bool IsConsecutiveBlock()
+            {
+                return originalBlock.ParentHash == _lastFinalizingBLockHash;
+            }
+
+            bool ConsecutiveBlockWillFinalizeBlocks()
+            {
+                _lastFinalizingBLockValidators.Add(block);
+                return _lastFinalizingBLockValidators.Count >= minSealersForFinalization;
+            }
+
+            List<BlockHeader> finalizedBlocks;
+            var isConsecutiveBlock = IsConsecutiveBlock();
+            
+            // Optimization:
+            // if block is consecutive than we can just check if this sealer will cause any blocks get finalized
+            // using cache of vallidators of not yet finalized blocks from previous block operation
+            if (isConsecutiveBlock && !ConsecutiveBlockWillFinalizeBlocks())
+            {
+                finalizedBlocks = Empty;
+            }
+            else
+            {
+                if (!isConsecutiveBlock)
+                {
+                    _lastFinalizingBLockValidators.Clear();
+                }
+
+                finalizedBlocks = new List<BlockHeader>();
+                var validators = new HashSet<Address>();
+                var originalBlockSealer = originalBlock.Beneficiary;
+
+                using (var batch = _blockInfoRepository.StartBatch())
+                {
+                    var (chainLevel, blockInfo) = GetBlockInfo(block);
+
+                    // Optimization:
+                    // if this block sealer seals for 2nd time than this seal can not finalize any blocks
+                    // as the 1st seal or some seal between 1st seal and current one would already finalize some of them
+                    bool OriginalBlockSealerSignedOnlyOnce() => !validators.Contains(originalBlockSealer) || block.Beneficiary != originalBlockSealer;
+                    
+                    while (!blockInfo.IsFinalized && OriginalBlockSealerSignedOnlyOnce())
+                    {
+                        validators.Add(block.Beneficiary);
+                        if (validators.Count >= minSealersForFinalization)
+                        {
+                            blockInfo.IsFinalized = true;
```

Comment:
so in the end it was required? we had this idea where we only had to store finalization levels?

Reply:
This simplifies things A LOT. Is fairly lightweight (1 byte per block) and might be used well in future scenarios:
1. Other consensus algorithms, even including ethash.
2. Checking if transaction is finalized from RPC ect.

Reply:
ok, convincing